File: python/pso.py
import json
from input import input
from encode import Encode
from decode import Decode
from particle import Particle
from ga import lsa
import traceback
import numpy as np
import math


def run_algorithm(commands, workers, pop_size, max_iter):
    (
        key_map,
        num_machines,
        num_workers,
        num_operations,
        num_jobs,
        num_operations_per_job,
        min_start_date_str,
        start_date_in_hours,
        avail_machines,
        avail_workers,
        lot_sizes,
        job_operations,
        job_lots,
        job_total_operations,
        job_operation_index,
        prev_operations,
        hprs,
        machines_avail_start_time,
        machines_avail_end_time,
        workers_avail_start_time,
        workers_avail_end_time,
    ) = input(commands, workers)

    print(commands, workers)

    lambda_theshold = 0.3
    min_sublot_size = 2000
    beta = 500

    decoder = Decode(
        key_map,
        num_machines,
        num_workers,
        num_operations,
        num_jobs,
        min_start_date_str,
        start_date_in_hours,
        job_operations,
        job_lots,
        job_total_operations,
        job_operation_index,
        prev_operations,
        hprs,
        machines_avail_start_time,
        machines_avail_end_time,
        workers_avail_start_time,
        workers_avail_end_time,
    )

    encoder = Encode(
        num_machines,
        num_jobs,
        num_operations,
        num_operations_per_job,
        avail_machines,
        avail_workers,
        lot_sizes,
        lambda_theshold,
        min_sublot_size,
        beta,
    )

    gbest_first_matrix = None
    gbest_second_matrix = None
    gbest_fitness = None
    c1 = 0.75
    c2 = 0.75
    w_start = 0.9
    w_end = 0.4

    pop = []

    try:
        for _ in range(pop_size):
            particle = Particle(
                num_jobs,
                num_operations,
                num_machines,
                num_operations_per_job,
                lot_sizes,
                hprs,
                avail_machines,
                min_sublot_size,
                beta,
                encoder,
                decoder,
            )
            pop.append(particle)

            if gbest_fitness == None or particle.pbest_fitness < gbest_fitness:
                gbest_first_matrix = np.copy(particle.pbest_first_matrix)
                gbest_second_matrix = np.copy(particle.pbest_second_matrix)
                gbest_fitness = particle.pbest_fitness

        for iter in range(max_iter):
            w = w_start - (w_start - w_end) * math.tan((iter / max_iter) * math.pi / 4)
            for particle in pop:
                particle.update_position(
                    gbest_first_matrix, gbest_second_matrix, w, c1, c2
                )

                if particle.pbest_fitness < gbest_fitness:
                    gbest_first_matrix = np.copy(particle.pbest_first_matrix)
                    gbest_second_matrix = np.copy(particle.pbest_second_matrix)
                    gbest_fitness = particle.pbest_fitness

            first_matrix_lsa = lsa(
                np.copy(gbest_first_matrix),
                np.copy(gbest_second_matrix),
                num_operations,
                hprs,
                avail_machines,
                min_sublot_size,
                beta,
                decoder,
            )

            gbest_first_matrix = np.copy(first_matrix_lsa)
            gbest_fitness = decoder.decode(gbest_first_matrix, gbest_second_matrix)

            # print("iter: ", iter, " gbest: ", gbest_fitness)

        result = decoder.decode(gbest_first_matrix, gbest_second_matrix, True)

        print("Final result: ", result)

        return result

    except Exception as e:
        print(f"Exception occurred: {e}")
        print(traceback.format_exc())
        raise e


from command_input import CommandInput
from worker_input import WorkerInput

if __name__ == "__main__":
    # Đọc dữ liệu từ file data.json với mã hóa UTF-8
    with open("data.json", "r", encoding="utf-8") as f:
        data = json.load(f)

    # Chuyển đổi commands và workers thành các đối tượng
    commands = [CommandInput(**cmd) for cmd in data["commands"]]
    workers = [WorkerInput(**worker) for worker in data["workers"]]
    pop_size = data["pop_size"]
    max_iter = data["max_iter"]

    # Chạy thuật toán
    result = run_algorithm(commands, workers, pop_size, max_iter)

    # Ghi kết quả vào file output.json
    with open("output.json", "w", encoding="utf-8") as f:
        json.dump(result, f, indent=4, ensure_ascii=False)

File: js/pso.js
// Import các thư viện cần thiết
const Particle = require("./particle"); // Đường dẫn đến file particle.js
const Encode = require("./encode"); // Đường dẫn đến file encode.js
const Decode = require("./decode"); // Đường dẫn đến file decode.js
const { input } = require("./input"); // Đường dẫn đến file input.js
const { lsa } = require("./ga"); // Đường dẫn đến file lsa.js
const math = require("mathjs"); // Thư viện math.js để xử lý ma trận

function runAlgorithm(commands, workers, popSize, maxIter) {
    try {
        // Giả lập hàm input để lấy dữ liệu đầu vào
        const {
            keyMap,
            numMachines,
            numWorkers,
            numOperations,
            numJobs,
            numOperationsPerJob,
            minStartDateStr,
            startDateInHours,
            availMachines,
            availWorkers,
            lotSizes,
            jobOperations,
            jobLots,
            jobTotalOperations,
            jobOperationIndex,
            prevOperations,
            hprs,
            machinesAvailStartTime,
            machinesAvailEndTime,
            workersAvailStartTime,
            workersAvailEndTime,
        } = input(commands, workers);

        const lambdaThreshold = 0.3;
        const minSublotSize = 2000;
        const beta = 500;

        // Khởi tạo encoder và decoder
        const decoder = new Decode(
            keyMap,
            numMachines,
            numWorkers,
            numOperations,
            numJobs,
            minStartDateStr,
            startDateInHours,
            jobOperations,
            jobLots,
            jobTotalOperations,
            jobOperationIndex,
            prevOperations,
            hprs,
            machinesAvailStartTime,
            machinesAvailEndTime,
            workersAvailStartTime,
            workersAvailEndTime
        );

        const encoder = new Encode(
            numMachines,
            numJobs,
            numOperations,
            numOperationsPerJob,
            availMachines,
            availWorkers,
            lotSizes,
            lambdaThreshold,
            minSublotSize,
            beta
        );

        // Biến lưu trữ global best
        let gbestFirstMatrix = null;
        let gbestSecondMatrix = null;
        let gbestFitness = null;

        const c1 = 0.75;
        const c2 = 0.75;
        const wStart = 0.9;
        const wEnd = 0.4;

        const pop = [];

        // Khởi tạo quần thể (population)
        for (let i = 0; i < popSize; i++) {
            const particle = new Particle(
                numJobs,
                numOperations,
                numMachines,
                numOperationsPerJob,
                lotSizes,
                hprs,
                availMachines,
                minSublotSize,
                beta,
                encoder,
                decoder
            );
            pop.push(particle);

            if (gbestFitness === null || particle.pbestFitness < gbestFitness) {
                gbestFirstMatrix = math.clone(particle.pbestFirstMatrix);
                gbestSecondMatrix = math.clone(particle.pbestSecondMatrix);
                gbestFitness = particle.pbestFitness;
            }
        }

        // Vòng lặp tối ưu hóa
        for (let iter = 0; iter < maxIter; iter++) {
            const w =
                wStart -
                (wStart - wEnd) * Math.tan(((iter / maxIter) * Math.PI) / 4);

            for (const particle of pop) {
                particle.updatePosition(
                    gbestFirstMatrix,
                    gbestSecondMatrix,
                    w,
                    c1,
                    c2
                );

                if (particle.pbestFitness < gbestFitness) {
                    gbestFirstMatrix = math.clone(particle.pbestFirstMatrix);
                    gbestSecondMatrix = math.clone(particle.pbestSecondMatrix);
                    gbestFitness = particle.pbestFitness;
                }
            }

            const firstMatrixLsa = lsa(
                math.clone(gbestFirstMatrix),
                math.clone(gbestSecondMatrix),
                numOperations,
                hprs,
                availMachines,
                minSublotSize,
                beta,
                decoder
            );

            gbestFirstMatrix = math.clone(firstMatrixLsa);
            gbestFitness = decoder.decode(gbestFirstMatrix, gbestSecondMatrix);
        }

        // Trả về kết quả cuối cùng
        const result = decoder.decode(
            gbestFirstMatrix,
            gbestSecondMatrix,
            true
        );

        return result;
    } catch (e) {
        console.error(`Exception occurred: ${e}`);
        throw e;
    }
}

const fs = require("fs"); // Thư viện để đọc file
const path = require("path"); // Thư viện để xử lý đường dẫn

// Đường dẫn đến file data.json
const filePath = path.join(__dirname, "data.json");

// Hàm đọc file và tạo các biến
function loadData() {
    try {
        // Đọc nội dung file
        const fileContent = fs.readFileSync(filePath, "utf8");

        // Parse nội dung file JSON
        const data = JSON.parse(fileContent);

        // Tạo các biến từ dữ liệu
        const commands = data.commands || [];
        const workers = data.workers || [];
        const popSize = data.pop_size || 10; // Giá trị mặc định là 10
        const maxIter = data.max_iter || 10; // Giá trị mặc định là 10

        // Trả về các biến
        return { commands, workers, popSize, maxIter };
    } catch (error) {
        console.error("Lỗi khi đọc file:", error);
        return null;
    }
}

// Gọi hàm loadData để đọc dữ liệu
const { commands, workers, popSize, maxIter } = loadData();

const r = runAlgorithm(commands, workers, popSize, maxIter);

console.log("Kết quả cuối cùng:", r);

fs.writeFileSync("output.json", JSON.stringify(r, null, 2), "utf8", (err) => {
    if (err) {
        console.error("Lỗi khi ghi file:", err);
    } else {
        console.log("Ghi file thành công!");
    }
});
